File: lib-rs/src/implementations/index_compress.rs
pub struct IndexCompress<T>
where
  T: Ord + Clone,
{
  compressed_vals: Vec<T>,
}

impl<T> IndexCompress<T>
where
  T: Ord + Clone,
{
  pub fn new(mut vec: Vec<T>) -> Self
  {
    vec.sort();
    vec.dedup();
    Self {
      compressed_vals: vec,
    }
  }

  pub fn get_compressed_index(&self, value: &T) -> usize
  {
    match self.compressed_vals.binary_search(value)
    {
      Ok(idx) | Err(idx) => idx,
    }
  }

  pub fn get_max_range(&self) -> usize
  {
    self.compressed_vals.len().saturating_sub(1)
  }
}

File: Library/cs/Collections/IndexCompress.cs
namespace Evermore.Collections;

using System;

public class IndexCompress<T> where T : IComparable<T>
{
    private readonly T[] _arr;
    private readonly int _size;

    public IndexCompress(T[] vec)
    {
        _arr = (T[])vec.Clone();
        Array.Sort(_arr);

        int unique = 0;
        if (_arr.Length > 0)
        {
            unique = 1;
            for (int i = 1; i < _arr.Length; i++)
            {
                if (_arr[i].CompareTo(_arr[i - 1]) != 0)
                {
                    _arr[unique++] = _arr[i];
                }
            }
        }

        _size = unique;
    }

    public int GetCompressedIndex(T value)
    {
        int index = Array.BinarySearch(_arr, 0, _size, value);
        return index >= 0 ? index : ~index;
    }

    public int GetMaxRange()
    {
        return _size - 1;
    }

    public T GetOriginalValue(int compressedIndex)
    {
        if (compressedIndex < 0 || compressedIndex >= _size)
            throw new ArgumentOutOfRangeException(nameof(compressedIndex));

        return _arr[compressedIndex];
    }
}
